keps/sig-storage/1710-selinux-relabeling/README.md

```diff
@@ -0,0 +1,477 @@
+It does so by running each container with unique *SELinux context* (such as `system_u:system_r:container_t:s0:c309,c383`) and labeling all content on all volumes with the corresponding label (`system_u:object_r:container_file_t:s0:c309,c383`).
+Only process with the context `...:container_t:s0:c309,c383` can access files with label `container_file_t:s0:c309,c383`, even if the process runs as root.
+Therefore rogue user cannot access potentially secret data of other containers, because volumes of each container have different label.
+
+In further text, we're going to shorten both `system_u:system_r:container_t:s0:c309,c383` (context of a process) and `system_u:object_r:container_file_t:s0:c309,c383` (label of a file) to `s0:309:383`.
+
+See [SELinux documentation](https://selinuxproject.org/page/NB_MLS) for more details.
+
+### SELinux context assignment
+In Kubernetes, the SELinux context of a pod is assigned in two ways:
+1. Either it is set by user in PodSpec or Container: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/.
+1. If not set in Pod/Container, the container runtime will allocate a new unique SELinux context and assign it to a pod (container) by itself.
+
+### Volumes
+Currently, Kubernetes *knows*(*) which volume plugins supports SELinux (i.e. supports extended attributes on a filesystem the plugin provides).
+If SELinux is supported for a volume, it passes the volume to the container runtime with ":Z" option ("private unshared").
+The container runtime then **recursively relabels** all files on the volume to either the label set in PodSpec/Container or the random value allocated by the container runtime itself.
+
+**This relabeling needs to traverse through the whole volume, and it can be slow for volumes with a large amount of files.**
+
+*) These in-tree volume plugins don't support SELinux: Azure File, CephFS, GlusterFS, HostPath, NFS, Portworx and Quobyte.
+All other volume plugins support it.
+This knowledge is hardcoded in in-tree volume plugins (e.g. [NFS](https://github.com/kubernetes/kubernetes/blob/0c5c3d8bb97d18a2a25977e92b3f7a49074c2ecb/pkg/volume/nfs/nfs.go#L235)).
+
+For CSI, kubelet uses following heuristics:
+
+1. Mount the volume (via `NodeStage` + `NodePublish` CSI calls).
+2. Check mount options of the volume mount dir. If and only if it contains `seclabel` mount option, the volume supports SELinux.
+
+### Goals
+
+Optionally (chosen by user), do not recursively relabel content of the volumes.
+
+### Non-Goals
+
+Change container runtimes / CRI.
+
+## Proposal
+
+Offer option in `Pod.Spec.PodSecurityContext to` *mount* volumes with the right labels instead of recursive relabeling:
+
+```go
+type SELinuxRelabelPolicy string
+
+const (
+    OnVolumeMount SELinuxRelabelPolicy = "OnVolumeMount"
+    AlwaysRelabel SELinuxRelabelPolicy = "Always"
```

Comment:
We can debate naming later, but I prefer something like "Recursive" instead of "Always", because we will relabel in all cases (if driver supports), regardless of what value is set here.